src/System.IO.Abstractions.TestingHelpers/MockFileStream.cs

```diff
@@ -60,7 +61,8 @@ public class MockFileStream : MemoryStream
                 }
 
                 fileData = new MockFileData(new byte[] { });
-                fileData.CreationTime = fileData.LastWriteTime = fileData.LastAccessTime = DateTime.Now;
+                mockFileDataAccessor.AdjustTimes(fileData,
+                    TimeAdjustments.CreationTime | TimeAdjustments.LastAccessTime);
```

Comment:
Is it intentional that we no longer update the `LastWriteTime` ?
I have to admit I haven't checked what the correct behavior is, but just noticed that this is different from what the code did before.

Reply:
No, this was not intentional, but correct all the same :-).
However, when I compared the behaviour to the actual behaviour of the real file system, I got some interesting results in the other case (that the file already exists).
The correct logic there would depend both on the `mode` and the `access`:
```csharp
        private TimeAdjustments GetTimeAdjustmentsForFileStreamWhenFileExists(FileMode mode, FileAccess access)
        {
            switch (mode)
            {
                case FileMode.Append:
                case FileMode.CreateNew:
                    if (access != FileAccess.Write)
                    {
                        return TimeAdjustments.LastAccessTime;
                    }
                    return TimeAdjustments.None;
                case FileMode.Create:
                case FileMode.Truncate:
                    if (access != FileAccess.Read)
                    {
                        return TimeAdjustments.LastAccessTime | TimeAdjustments.LastWriteTime;
                    }
                    return TimeAdjustments.LastAccessTime;
                case FileMode.Open:
                case FileMode.OpenOrCreate:
                default:
                    return TimeAdjustments.None;
            }
        }
```

Reply:
For reference:
This was my system test to compare the behaviour:
Should/Can I include such tests? Would it make sense to include this kind of tests e.g. in a separate test suite?

```csharp
using System.Collections.Generic;
using NUnit.Framework;
using System.Threading;
using System.Linq;

namespace System.IO.Abstractions.TestingHelpers.Tests
{
    [TestFixture]
    public class SystemTests
    {
        public static IEnumerable<object[]> GetSystemTestParameters()
        {
            foreach (var fileMode in Enum.GetValues(typeof(FileMode)).Cast<FileMode>())
            {
                foreach (var fileAccess in Enum.GetValues(typeof(FileAccess)).Cast<FileAccess>())
                {
                    yield return new object[] { true, fileMode, fileAccess };
                    yield return new object[] { false, fileMode, fileAccess };
                }
            }
        }

        [Test, TestCaseSource(nameof(GetSystemTestParameters))]
        public void Stream_Create_ShouldBehaveSameAsRealFileSystem(bool fileExists, FileMode fileMode, FileAccess fileAccess)
        {
            string fileName = @$"c:\test\existing_{fileExists}_{fileMode}_{fileAccess}.txt";
            if (fileExists)
            {
                File.WriteAllText(fileName, "foo");
            }
            else
            {
                File.Delete(fileName);
            }

            var time_before = DateTime.UtcNow.AddDays(-5);
            var time_after = time_before.AddSeconds(1);
            var fs = new MockFileSystem().MockTime(() => time_before);
            fs.Directory.CreateDirectory(@$"c:\test");
            if (fileExists)
            {
                fs.File.WriteAllText(fileName, "foo");
            }

            var realfilesystem_before_accessed = File.GetLastAccessTimeUtc(fileName);
            var realfilesystem_before_written = File.GetLastWriteTimeUtc(fileName);
            var realfilesystem_before_created = File.GetCreationTimeUtc(fileName);
            Thread.Sleep(1000);
            try
            {
                _ = new FileStream(fileName, fileMode, fileAccess);
            }
            catch (Exception ex)
            {
                fs.MockTime(() => time_after);
                Assert.Throws(ex.GetType(), () =>
                {
                    _ = fs.FileStream.Create(fileName, fileMode, fileAccess);
                });
                return;
            }

            var realfilesystem_after_accessed = File.GetLastAccessTimeUtc(fileName);
            var realfilesystem_after_written = File.GetLastWriteTimeUtc(fileName);
            var realfilesystem_after_created = File.GetCreationTimeUtc(fileName);

            bool isRealCreatedChanged = realfilesystem_before_created != realfilesystem_after_created;
            bool isRealAccessedChanged = realfilesystem_before_accessed != realfilesystem_after_accessed;
            bool isRealWrittenChanged = realfilesystem_before_written != realfilesystem_after_written;

            var mockfilesystem_before_accessed = fs.File.GetLastAccessTimeUtc(fileName);
            var mockfilesystem_before_written = fs.File.GetLastWriteTimeUtc(fileName);
            var mockfilesystem_before_created = fs.File.GetCreationTimeUtc(fileName);
            fs.MockTime(() => time_after);
            _ = fs.FileStream.Create(fileName, fileMode, fileAccess);
            var mockfilesystem_after_accessed = fs.File.GetLastAccessTimeUtc(fileName);
            var mockfilesystem_after_written = fs.File.GetLastWriteTimeUtc(fileName);
            var mockfilesystem_after_created = fs.File.GetCreationTimeUtc(fileName);

            if (isRealCreatedChanged)
            {
                Assert.That(mockfilesystem_before_created, Is.Not.EqualTo(mockfilesystem_after_created), message: "Created");
            }
            else
            {
                Assert.That(mockfilesystem_before_created, Is.EqualTo(mockfilesystem_after_created), message: "Created");
            }
            if (isRealAccessedChanged)
            {
                Assert.That(mockfilesystem_before_accessed, Is.Not.EqualTo(mockfilesystem_after_accessed), message: "Accessed");
            }
            else
            {
                Assert.That(mockfilesystem_before_accessed, Is.EqualTo(mockfilesystem_after_accessed), message: "Accessed");
            }
            if (isRealWrittenChanged)
            {
                Assert.That(mockfilesystem_before_written, Is.Not.EqualTo(mockfilesystem_after_written), message: "Written");
            }
            else
            {
                Assert.That(mockfilesystem_before_written, Is.EqualTo(mockfilesystem_after_written), message: "Written");
            }
        }
    }
}
```

I realized, that in the following cases the `MockFileSystem` does not throw any or the correct exception in comparison with the real file system:
![image](https://user-images.githubusercontent.com/3438234/188125522-90e6271d-df01-46e0-a1ca-a78cdbfc34f3.png)
But for this I will create a separate issue and not fix it with this PR!

Reply:
Generally I like the idea of having functional parity tests, but I think it would be quite hard to maintain and overlap a lot with the existing unit tests. Also hard for contributors - would we expect them to add unit and system tests for each feature/fix? What's the criteria for when to pick what?

Reply:
I agree that it would be difficult to maintain. That's why I didn't include it yet, but it helped me during the implementation.
Sometimes this kind of tests might be beneficial but I would keep them separate from unit tests and optional.

Reply:
Good arguments. Nevertheless, I fear that this time handling is quite difficult to get right, especially as the documentation is quite lacking when the time values are updated :-(